util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -74,195 +93,221 @@ var (
-	klog.Info("Starting TCP Server")
-	w.startWork(rw, request, ProtocolTCP, "iperf", tcpServerArguments)
-}
-
-// StartUDPServer starts iperf server for udp measurements.
-func (w *Worker) StartUDPServer(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting UDP Server")
-	w.startWork(rw, request, ProtocolUDP, "iperf", udpServerArguments)
-}
-
-// StartHTTPClient starts an http client for http measurements.
-func (w *Worker) StartHTTPClient(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting HTTP Client")
-	w.startWork(rw, request, ProtocolHTTP, "siege", httpClientArguments)
+func (w *Worker) populateResourceSpec(object runtime.Object) error {
+	var err error
+	w.work.resourceContent, err = runtime.DefaultUnstructuredConverter.ToUnstructured(object)
+	if err != nil {
+		return fmt.Errorf("error converting event Object to unstructured.Event object: %s", object)
+	}
+	w.work.resourceSpecMap = w.work.resourceContent["spec"].(map[string]interface{})
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(w.work.resourceSpecMap, &w.work.resourceSpec); err != nil {
+		return fmt.Errorf("error converting custom resource properties: %s", err)
+	}
+	return nil
 }
 
-// StartTCPClient starts iperf client for tcp measurements.
-func (w *Worker) StartTCPClient(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting TCP Client")
-	w.startWork(rw, request, ProtocolTCP, "iperf", tcpClientArguments)
+func (w *Worker) startWork() {
+	if w.work.resourceSpec.Protocol == ProtocolHTTP && w.work.workType == "server" {
+		w.StartHTTPServer()
+		return
+	}
+	var err error
+	w.work.arguments, err = populateTemplates(protocolArgumentMap[w.work.workType][w.work.resourceSpec.Protocol], w.work.resourceSpecMap)
+	if err != nil {
+		err = fmt.Errorf("populating template failed: %v", err)
+		klog.Error(err)
+		w.updateStatus(makeErrorStatus(err))
+		return
+	}
+	klog.Infof("Populated templates: %s", w.work.arguments)
+	startTimestamp := w.getStartTimestamp()
+	w.schedule(startTimestamp)
 }
 
-// StartUDPClient starts iperf client for udp measurements.
-func (w *Worker) StartUDPClient(rw http.ResponseWriter, request *http.Request) {
-	klog.Info("Starting UDP Client")
-	w.startWork(rw, request, ProtocolUDP, "iperf", udpClientArguments)
+func (w *Worker) getStartTimestamp() int64 {
+	if w.work.workType == "client" {
+		return w.work.resourceSpec.ClientStartTimestamp
+	}
+	return time.Now().Unix()
 }
 
-// Handler handles http requests for http measurements.
-func (w *Worker) Handler(rw http.ResponseWriter, request *http.Request) {
-	w.sendResponse(rw, http.StatusOK, "ok")
+func (w *Worker) updateStatus(status map[string]interface{}) {
+	w.work.resourceContent["status"] = status
+	var unstructuredRes unstructured.Unstructured
+	unstructuredRes.SetUnstructuredContent(w.work.resourceContent)
+	_, err := w.resourceInterface.Update(context.TODO(), &unstructuredRes, metav1.UpdateOptions{})
```

Comment:
Please add a TODO to add retries (similar to wht we do in: https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/framework/client/objects.go#L276 )

Reply:
Added.

Reply:
We generally try to avoid owner-less TODOs, so either please add yourself (i.e. TODO(@VivekThrivikraman-est): ...) or open an issue for that and put issue number

Reply:
Modified.